docs/madr/decisions/010-timeout-policy.md

```diff
@@ -0,0 +1,174 @@
+
+TBD.
+
+## Solution
+
+### How Timeout policy can be configured right now
+This is the default timeout configuration in Kuma right now:
+
+```yaml
+ conf:
+    connectTimeout: 5s # all protocols
+    tcp: 
+      idleTimeout: 1h 
+    http:
+      requestTimeout: 15s 
+      idleTimeout: 1h
+      streamIdleTimeout: 30m
+      maxStreamDuration: 0s
+```
+
+Why idle timeout can be configured on both TCP and HTTP connections?
+Is there a chance that client would like to configure different idle timeouts for database/Kafka and for HTTP services? - I can't come up on any valid case when this can be used.
+
+### What timeouts can we configure in envoy
+1. TCP specific
+    1. connect timeout
+    2. idle timeout
+2. HTTP/GRPC connections specific
+    1. connection timeouts
+        1. idle timeout
+        2. max connection duration
+    2. stream timeouts
+        1. request timeout
+        2. request headers timeout
+        3. stream idle timeout
+        4. max stream duration
+    3. route timeout
+        1. idle timeout
+        2. timeout
+        3. per try timeout
+        4. per try idle timeout
+        5. max stream duration
+           Detailed explanation of those timeouts can be found in [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/faq/configuration/timeouts).
+
+#### Common timeouts
+As we can see, `connection` and `idle timeout` can be configured for every connection type. There are two options here:
+
+1. Keep separate idle timeouts for TCP and HTTP
+2. Merge those two timeouts into single configuration
+
+Ad 1. It gives clients more fine-tuning of timeouts, but it can cause chaos
+Ad 2. Simpler config and you can still achieve the same result as previously by configuring it for `MeshSubset|MeshService|MeshServiceSubset`
+
+#### Connection and stream timeout
+I understand that `max connection timeout` and `max stream timeout ` are basically doing the same thing, since every 
+connection event in HTTP 1 are treated as streams in Envoy according to the docs.
+
+#### Route timeouts
+As for the route timeouts, we only care about `idle timeout` and `timeout` since `per try timeout` and `per try idle timeout` 
+are retry related timeouts and should be configured in retry policy.
+
+#### Request headers timeout
+Last one specific timeout is `request headers timeout` but this timeout makes no sense on outbound connections. 
+This only can be used to protect yourself from malicious inbound clients.
+
+#### Scaled timeouts
+Moreover, there is a mechanism of scaled timeout in envoy, which I am not sure if we want to expose to our users, at least not at first.
+
+### Specification
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshUpstreamTimeout is outbound policy, that's why it has a "to" array:
+```yaml
+to:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshHTTPRoute
```

Comment:
We cannot use `MeshSubset` and `MeshServiceSubset` here. Not sure about `MeshHTTPRoute` (check 
 what was done in MeshAccessLog). We have the same limitation today with the current Timeout policy and this will be a limitation for every policy with `to`.

Because we build timeout on outbound listener, we don't know what specific instance will be picked for this listener.

Reply:
Oh, you are right, I think I misunderstood how `MeshSubset` and `MeshServiceSubset` works. I will update madr accordingly